### PR TITLE
POLIO-1237 [front-end] multi round budget tab 

### PR DESCRIPTION
**Pull request by beygorghor**
[Mock-ups](https://whimsical.com/budget-tab-move-QyBCGa5c54bTxYb6Wed1gN)

Move budget information from campaign dialog to budget process dialog.

Related JIRA tickets :POLIO-1237

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [x] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes

- adapt serializers to return and validate round dict
- adding target_population to rounds serializer
- move all fields to edit and create modale

## How to test

Go to budget and create /edit a process, play with all the fields and save.

## Print screen / video

https://github.com/BLSQ/iaso/assets/12494624/79fda1a5-f531-49e9-b317-91a86e9708e6

### CODE REVIEW

File: plugins/polio/budget/serializers.py [hunk 1]
```diff
@@ -1,4 +1,6 @@
+from collections import OrderedDict
 from enum import Enum
+from pprint import pformat
```
Comment (kemar) [inline plugins/polio/budget/serializers.py]: This could be removed.

File: plugins/polio/budget/serializers.py [hunk 2]
```diff
@@ -79,13 +82,34 @@ class AvailableRoundsSerializer(serializers.Serializer):
...
+        model = Round
+        fields = [
+            "id",
+            "number",
+            "cost",
+            "target_population",
+        ]
+
+
+class BudgetProcessUWriteRoundSerializer(serializers.ModelSerializer):
```
Comment (kemar) [inline plugins/polio/budget/serializers.py]:
```suggestion
class BudgetProcessWriteRoundSerializer(serializers.ModelSerializer):
```

File: plugins/polio/budget/serializers.py [hunk 3]
```diff
@@ -79,13 +82,34 @@ class AvailableRoundsSerializer(serializers.Serializer):
...
+    class Meta:
+        model = Round
+        fields = [
+            "id",
+            "cost",
```
Comment (kemar) [inline plugins/polio/budget/serializers.py:100]: It's terrifying how the addition of a _single_ field in a nested writable serializer can generate so much extra code in the parent validation logic…

There should be a simpler solution, but I can't find it…
So I've got nothing better to suggest.

File: plugins/polio/budget/serializers.py [hunk 4]
```diff
@@ -129,60 +153,65 @@ class Meta:
...
+        self.handle_rounds(budget_process, rounds_data)
         return budget_process
 
     def update(self, budget_process: BudgetProcess, validated_data: dict) -> BudgetProcess:
+        rounds_data = validated_data.pop("rounds", [])
+
         budget_process = super().update(budget_process, validated_data)
-        # Unlink old rounds.
-        budget_process.rounds.update(budget_process=None)
-        # Link new rounds.
```
Comment (kemar) [inline plugins/polio/budget/serializers.py:171]: These comments added a little bit of explanation: "Unlink old rounds", then "Link new rounds".

Tt would have been nice to leave them ;)

File: plugins/polio/budget/serializers.py [hunk 5]
```diff
@@ -129,60 +153,65 @@ class Meta:
...
+        for round_data in rounds_data:
+            round_id = round_data.get("id")
+            if round_id:
+                round_instance = Round.objects.get(id=round_id)
+                round_serializer = BudgetProcessUWriteRoundSerializer(round_instance, data=round_data, partial=True)
+                if round_serializer.is_valid():
+                    round_serializer.save(budget_process=budget_process)
 
-        # Create a new `BudgetProcess`.
+    def create(self, validated_data):
```
Comment (kemar) [inline plugins/polio/budget/serializers.py]:
```suggestion
    def create(self, validated_data: dict) -> BudgetProcess:
```

File: plugins/polio/budget/serializers.py [hunk 6]
```diff
@@ -129,60 +153,65 @@ class Meta:
...
-        budget_process.rounds.update(budget_process=None)
-        # Link new rounds.
-        rounds_ids = [round.id for round in self.validated_data["rounds"]]
-        Round.objects.filter(id__in=rounds_ids).update(budget_process=budget_process)
+        existing_round_ids = set(budget_process.rounds.values_list("id", flat=True))
+        new_round_ids = set(round_data["id"] for round_data in rounds_data if "id" in round_data)
+
+        # should we also empty cost?
+        rounds_to_unlink = existing_round_ids - new_round_ids
+        if rounds_to_unlink:
```
Comment (kemar) [inline plugins/polio/budget/serializers.py:209]: 👍

File: plugins/polio/js/src/domains/Budget/BudgetProcess/BudgetProcessApproval.tsx [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        useFormikContext<BudgetDetail>();
+    const updateBudgetStatus = useCallback(
+        (fieldName: string, value: any) => {
+            const fieldKey = fieldName.replace(WORKFLOW_SUFFIX, '');
+            const computedBudgetStatusIndex = findBudgetStateIndex(values);
+            const fieldIndex = BUDGET_STATES.findIndex(
+                budgetState => budgetState === fieldKey,
+            );
+            if (fieldIndex > computedBudgetStatusIndex && value) {
+                // setFieldValue('budget_status', fieldKey);
```
Comment (kemar) [inline plugins/polio/js/src/domains/Budget/BudgetProcess/BudgetProcessApproval.tsx]: Comments could probably be deleted.

File: plugins/polio/budget/serializers.py [hunk 8]
```diff
@@ -129,60 +153,65 @@ class Meta:
...
+        rounds_campaigns = {round.campaign_id for round in rounds}
         if len(rounds_campaigns) > 1:
-            raise serializers.ValidationError(f"Rounds must be from the same campaign: {submitted_rounds}.")
+            raise serializers.ValidationError("Rounds must be from the same campaign.")
 
         return submitted_rounds
 
-    def create(self, validated_data: dict) -> BudgetProcess:
-        request = self.context["request"]
+    def handle_rounds(self, budget_process, rounds_data):
```
Comment (kemar) [inline plugins/polio/budget/serializers.py]:
```suggestion
    def handle_rounds(self, budget_process: BudgetProcess, rounds_data: dict) -> None:
```

Let's be consistent :)

**Review by kemar — APPROVED** [review]
Backend LGTM

I suggest some changes but only if you're OK with it. Nothing mandatory.

File: plugins/polio/js/src/domains/Budget/BudgetProcess/BudgetProcessCostPerChild.tsx [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        fontWeight: 'bold',
+        textAlign: 'right',
+    },
+    noRounds: {
+        textAlign: 'center',
+    },
+};
+
+const currency = '$';
+const placeholder = '--';
```
Comment (quang-le) [inline plugins/polio/js/src/domains/Budget/BudgetProcess/BudgetProcessCostPerChild.tsx:75]: import `textPlaceholder` from bluesquare-components

File: plugins/polio/js/src/domains/Budget/BudgetProcess/BudgetProcessCostPerChild.tsx [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+                })}
+                <TableRow>
+                    <TableCell colSpan={2} />
+                    <TableCell sx={styles.totalTitle}>
+                        {formatMessage(MESSAGES.costPerChildTotal)}:
+                    </TableCell>
+                    <TableCell sx={styles.total}>{totalCostPerChild}</TableCell>
+                </TableRow>
+            </TableBody>
+        </Table>
```
Comment (quang-le) [inline plugins/polio/js/src/domains/Budget/BudgetProcess/BudgetProcessCostPerChild.tsx:180]: That's a component

File: plugins/polio/js/src/domains/Budget/BudgetProcess/BudgetProcessRelease.tsx [hunk 11]
```diff
@@ -0,0 +1,76 @@
+import { Box, Grid } from '@mui/material';
+import { Field, useFormikContext } from 'formik';
+import React, { FunctionComponent } from 'react';
+
+import { NumberInput, useSafeIntl } from 'bluesquare-components';
+import { DateInput, PaymentField } from '../../../components/Inputs';
+import MESSAGES from '../messages';
+import { BudgetDetail } from '../types';
+
+export const BudgetProcessRelease: FunctionComponent = () => {
```
Comment (quang-le) [inline plugins/polio/js/src/domains/Budget/BudgetProcess/BudgetProcessRelease.tsx]: BudgetProcessFundsRelease

File: plugins/polio/js/src/domains/Budget/BudgetProcess/CreateBudgetProcessModal.tsx [hunk 12]
```diff
@@ -86,31 +89,39 @@ const CreateBudgetProcessModal: FunctionComponent<Props> = ({
...
+                                options={currentCampaignOptions}
+                            />
+                        </Grid>
+                        <Grid item xs={6}>
+                            <Field
+                                label={formatMessage(MESSAGES.labelRound)}
+                                name="rounds"
+                                component={MultiSelect}
+                                options={currentRoundsOptions}
+                                returnFullObject
```
Comment (quang-le) [inline plugins/polio/js/src/domains/Budget/BudgetProcess/CreateBudgetProcessModal.tsx:120]: I managed to crash the modal while playing with rounds. I think it maybe when you have this prop to `true` and the value is `null`

File: plugins/polio/js/src/domains/Budget/BudgetProcess/validation.ts [hunk 13]
```diff
@@ -1,37 +1,215 @@
...
+            .optional()
+            .matches(
+                /^\d{4}-\d{2}-\d{2}$/,
+                formatMessage(MESSAGES.invalidDate),
+            ),
+        approved_by_unicef_at_WFEDITABLE: string()
+            .nullable()
+            .optional()
+            .matches(
+                /^\d{4}-\d{2}-\d{2}$/,
```
Comment (quang-le) [inline plugins/polio/js/src/domains/Budget/BudgetProcess/validation.ts]: You should save this regex somewhere in a constant either in iaso or bluesquare-components

**Review by quang-le — CHANGES_REQUESTED** [review]
There's this crash:
https://github.com/BLSQ/iaso/assets/38907762/e4391945-66c5-4a45-ba34-8113c1321120

**Pull request merged**